sympy/solvers/ode/single.py

```diff
@@ -382,6 +381,132 @@ def inverse(self):
+    \partial{}F/\partial{}x` and `Q(x, y) = \partial{}F/\partial{}y`.  It can
+    be shown that a necessary and sufficient condition for a first order ODE
+    to be exact is that `\partial{}P/\partial{}y = \partial{}Q/\partial{}x`.
+    Then, the solution will be as given below::
+
+        >>> from sympy import Function, Eq, Integral, symbols, pprint
+        >>> x, y, t, x0, y0, C1= symbols('x,y,t,x0,y0,C1')
+        >>> P, Q, F= map(Function, ['P', 'Q', 'F'])
+        >>> pprint(Eq(Eq(F(x, y), Integral(P(t, y), (t, x0, x)) +
+        ... Integral(Q(x0, t), (t, y0, y))), C1))
+                    x                y
+                    /                /
+                   |                |
+        F(x, y) =  |  P(t, y) dt +  |  Q(x0, t) dt = C1
+                   |                |
+                  /                /
+                  x0               y0
+
+    Where the first partials of `P` and `Q` exist and are continuous in a
+    simply connected region.
+
+    A note: SymPy currently has no way to represent inert substitution on an
+    expression, so the hint ``1st_exact_Integral`` will return an integral
+    with `dy`.  This is supposed to represent the function that you are
+    solving for.
+
+    Examples
+    ========
+
+    >>> from sympy import Function, dsolve, cos, sin
+    >>> from sympy.abc import x
+    >>> f = Function('f')
+    >>> dsolve(cos(f(x)) - (x*sin(f(x)) - f(x)**2)*f(x).diff(x),
+    ... f(x), hint='1st_exact')
+    Eq(x*cos(f(x)) + f(x)**3/3, C1)
+
+    References
+    ==========
+
+    - https://en.wikipedia.org/wiki/Exact_differential_equation
+    - M. Tenenbaum & H. Pollard, "Ordinary Differential Equations",
+      Dover 1963, pp. 73
+
+    # indirect doctest
+
+    """
+    hint = "1st_exact"
+    has_integral = True
+    order = [1]
+
+    def _wilds(self, f, x, order):
+        P = Wild('P', exclude=[f(x).diff(x)])
+        Q = Wild('Q', exclude=[f(x).diff(x)])
+        return P, Q
+
+    def _equation(self, fx, x, order):
+        P, Q = self.wilds()
+        return P + Q*fx.diff(x)
+
+    def _verify(self, fx) -> bool:
+        P, Q = self.wilds()
+        x = self.ode_problem.sym
+        y = Dummy('y')
+
+        m, n = self.wilds_match()
+
+        m = m.subs(fx, y)
+        n = n.subs(fx, y)
+        numerator = cancel(m.diff(y) - n.diff(x))
+        if numerator.is_zero:
+            # Is exact
+            return True
+        else:
+            # The following few conditions try to convert a non-exact
+            # differential equation into an exact one.
+            # References:
+            # 1. Differential equations with applications
+            # and historical notes - George E. Simmons
+            # 2. https://math.okstate.edu/people/binegar/2233-S99/2233-l12.pdf
+
+            factor_n = cancel(numerator/n)
+            factor_m = cancel(-numerator/m)
+            if factor_n.free_symbols == {x}:
+                # If (dP/dy - dQ/dx) / Q = f(x)
```

Comment:
I guess this should be `if y not in factor_n.free_symbols`. Then it would work for an equation that has other symbols in it.

Reply:
Yes, you are correct. The current implementation didn't consider even the simple case of a constant non zero factor, such as in `2*x*f(x)*f(x).diff(x) + (1+x)*f(x)**2 - exp(x)`. This is fixed now. It also was not considering an ODE with some other symbol. For instance, `x*f(x).diff(x) + a*f(x) + x**2`, which is now is correctly identified as exact. But I'm struggling to understand why the result of dsolve doesn't work for this example:
```
In [3]: eq=x*f(x).diff(x) + a*f(x) + x**2

In [4]: dsolve(eq, hint='1st_linear')
Out[4]: 
       ⎛     ⎛⎧ 2  a⋅log(x)            ⎞⎞           
       ⎜     ⎜⎪x ⋅ℯ                    ⎟⎟           
       ⎜     ⎜⎪────────────  for a ≠ -2⎟⎟  -a⋅log(x)
f(x) = ⎜C₁ - ⎜⎨   a + 2                ⎟⎟⋅ℯ         
       ⎜     ⎜⎪                        ⎟⎟           
       ⎜     ⎜⎪   log(x)     otherwise ⎟⎟           
       ⎝     ⎝⎩                        ⎠⎠           

In [5]: dsolve(eq, hint='1st_exact')
Out[5]: 
⎡                                                     ⎧                                                                                                                                     2  a⋅
⎢              ⎧    2    2                            ⎪                          C₁⋅a                                                   2⋅C₁                                               x ⋅ℯ  
⎢              ⎪C₁⋅x  - x ⋅log(x)  for a = -2         ⎪- ──────────────────────────────────────────────────── - ──────────────────────────────────────────────────── + ──────────────────────────
⎢False, f(x) = ⎨                             , f(x) = ⎨   2               a⋅log(x)                   a⋅log(x)    2               a⋅log(x)                   a⋅log(x)    2               a⋅log(x) 
⎢              ⎪       nan         otherwise          ⎪  a ⋅log(x) - 2⋅a⋅ℯ         + 2⋅a⋅log(x) - 4⋅ℯ           a ⋅log(x) - 2⋅a⋅ℯ         + 2⋅a⋅log(x) - 4⋅ℯ           a ⋅log(x) - 2⋅a⋅ℯ         
⎢              ⎩                                      ⎪                                                                                                                                          
⎣                                                     ⎩                                                                                nan                                                       

log(x)                                        ⎧                                                                2  a⋅log(x)                                                                  ⎤
                                              ⎪           C₁⋅a                        2⋅C₁                    x ⋅ℯ                                                                          ⎥
──────────────────────────  for a = 0         ⎪───────────────────────── + ───────────────────────── - ─────────────────────────  for (a ≥ -∞ ∧ a < -2) ∨ (a > -2 ∧ a < 0) ∨ (a > 0 ∧ a < ∞)⎥
                  a⋅log(x)           , f(x) = ⎨   a⋅log(x)      a⋅log(x)      a⋅log(x)      a⋅log(x)      a⋅log(x)      a⋅log(x)                                                            ⎥
+ 2⋅a⋅log(x) - 4⋅ℯ                            ⎪a⋅ℯ         + 2⋅ℯ           a⋅ℯ         + 2⋅ℯ           a⋅ℯ         + 2⋅ℯ                                                                    ⎥
                                              ⎪                                                                                                                                             ⎥
                            otherwise         ⎩                                       nan                                                                 otherwise                         ⎦


```
The hints `1st_exact_Integral` and `1st_linear_Integral` return equivalent results, so it seems there is something weird happening in the integral evaluation, which I'm still trying to figure out.

Reply:
Does `get_general_solution` return the same result for both solvers?

Reply:
For the `1st_linear`, `_get_general_solution` returns:
```
Eq(f(x), (C1 + Integral(-x*exp(Integral(a/x, x)), x))*exp(-Integral(a/x, x)))
```
`1st_exact`:
```
Eq(Subs(Integral((_y*a + x**2)*exp(Integral((a - 1)/x, x)), x) + Integral(x*exp(Integral((a - 1)/x, x)) - Integral(a*exp(Integral((a - 1)/x, x)), x), _y), _y, f(x)), C1)
```

I say they are equivalent in the sense that if you solve both by hand, they evaluate to the same thing, which is the result of `dsolve(eq, hint='1st_linear')`.

Reply:
The error for this function is related to the order of operations. What we want is (using `m` from the example equation from above)
```
In [1]: a = Symbol('a', real = True)

In [2]: m = (a*y + x**2)*exp((a - 1)*log(x))

In [3]: Integral(m, x).doit().diff(y)
Out[3]: 
⎧   a⋅log(x)      a⋅log(x)            
⎪a⋅ℯ           2⋅ℯ                    
⎪─────────── + ───────────  for a ≠ -2
⎪   a + 2         a + 2               
⎨                                     
⎪           1                         
⎪           ──              otherwise 
⎪            2                        
⎩           x                         
```

But since the integral is not evaluated in `_get_general_solution` (`doit()` is called later), what we actually get is the derivative evaluated first, resulting in the wrong solution:
```
In [4]: Integral(m, x).diff(y).doit()
Out[4]: 
  ⎛⎧ a⋅log(x)           ⎞
  ⎜⎪ℯ                   ⎟
  ⎜⎪─────────  for a ≠ 0⎟
a⋅⎜⎨    a               ⎟
  ⎜⎪                    ⎟
  ⎜⎪ log(x)    otherwise⎟
  ⎝⎩                    ⎠
```
Is there a way to force the derivative evaluation after the integral?

Reply:
Why does the order make a difference?

The unevaluated `Derivative` could be used instead of `diff`:
```python
In [8]: Derivative(Integral(m, x), y).doit()                                                                                      
Out[8]: 
⎧   a⋅log(x)      a⋅log(x)            
⎪a⋅ℯ           2⋅ℯ                    
⎪─────────── + ───────────  for a ≠ -2
⎪   a + 2         a + 2               
⎨                                     
⎪           1                         
⎪           ──              otherwise 
⎪            2                        
⎩           x  
```

Reply:
At first, I thought this equation was violating some condition of the theorem which allows you to change the order and tried to find out what it was, but the issue seems to be in the integration itself:
```
In [1]: a = Symbol('a', real = True)

In [2]: m = (a*y + x**2)*exp((a - 1)*log(x))

In [3]: Integral(m, x).doit()
Out[3]:
⎧     a⋅log(x)    2  a⋅log(x)        a⋅log(x)
⎪a⋅y⋅ℯ           x ⋅ℯ             2⋅y⋅ℯ
⎪───────────── + ──────────── + ─────────────  for a ≠ -2
⎪    a + 2          a + 2           a + 2
⎨
⎪                         y
⎪                log(x) + ──                   otherwise
⎪                          2
⎩                         x

```
Now, distributing the `(a*y + x**2)` terms:
```
In [4]: Integral(m.expand(), x).doit()
Out[4]:

    ⎛⎧ a⋅log(x)           ⎞   ⎛⎧ 2  a⋅log(x)            ⎞
    ⎜⎪ℯ                   ⎟   ⎜⎪x ⋅ℯ                    ⎟
    ⎜⎪─────────  for a ≠ 0⎟   ⎜⎪────────────  for a ≠ -2⎟
 a⋅y⋅⎜⎨    a               ⎟ + ⎜⎨   a + 2                ⎟
    ⎜⎪                    ⎟   ⎜⎪                        ⎟
    ⎜⎪ log(x)    otherwise⎟   ⎜⎪   log(x)     otherwise ⎟
    ⎝⎩                    ⎠   ⎝⎩                        ⎠

```
I don't know why in the expanded case that `a!=0` shows up, while in the first only the `a!=-2` condition is used, since we are integrating the same expression. This second result is what generates that `a!=0`. I thought it was a result of changing the order, but this is incorrect. It's actually just a consequence of the different results from the integration:
```
In [5]: Integral(m, x).doit().diff(y)
Out[5]:
⎧   a⋅log(x)      a⋅log(x)
⎪a⋅ℯ           2⋅ℯ
⎪─────────── + ───────────  for a ≠ -2
⎪   a + 2         a + 2
⎨
⎪           1
⎪           ──              otherwise
⎪            2
⎩           x

In [6]: Integral(m.expand(), x).doit().diff(y)
Out[6]:

  ⎛⎧ a⋅log(x)           ⎞
  ⎜⎪ℯ                   ⎟
  ⎜⎪─────────  for a ≠ 0⎟
a⋅⎜⎨    a               ⎟
  ⎜⎪                    ⎟
  ⎜⎪ log(x)    otherwise⎟
  ⎝⎩                    ⎠

```
I suppose that using the derivative first expands `m` at some point, resulting in the mentioned result from `Integral(m, x).diff(y).doit()`. In any case, this doesn't seem directly related to the Exact ODE solving. This issue just didn't show up before in this context because the matching algorithm wasn't considering more than a single symbol. I believe this should be analyzed as a separate bug.

Reply:
Yes, that looks like a separate bug. Can you open an issue explaining it?

Putting that aside is it possible to fix the exact solver?

Reply:
Yes, I will open the issue.
I believe your suggestion of using `Derivative(Integral(m, x), y).doit()` helps. We get the correct solution, although the printing is not very clear since the integral generate those two conditions on the value of `a`: you must ignore the`nan` in each of the solutions to understand the result:
```python
In [1]: a = Symbol('a', real = True)

In [2]: eq = x*f(x).diff(x) + a*f(x) + x**2

In [3]: dsolve(eq, hint='1st_linear')
Out[3]:
       ⎛     ⎛⎧ 2  a⋅log(x)            ⎞⎞
       ⎜     ⎜⎪x ⋅ℯ                     ⎟⎟
       ⎜     ⎜⎪────────────  for a ≠ -2⎟⎟  -a⋅log(x)
f(x) = ⎜C₁ - ⎜⎨   a + 2                ⎟⎟⋅ℯ
       ⎜     ⎜⎪                        ⎟⎟
       ⎜     ⎜⎪   log(x)     otherwise ⎟⎟
       ⎝     ⎝⎩                        ⎠⎠

In [4]: dsolve(eq, hint='1st_exact')
Out[4]:
⎡                                                              ⎧                                                                2  a⋅log(x)                  ⎤
⎢       ⎧    -a⋅log(x)    -a⋅log(x)                            ⎪           C₁⋅a                        2⋅C₁                    x ⋅ℯ                          ⎥
⎢       ⎪C₁⋅ℯ          - ℯ         ⋅log(x)  for a = -2         ⎪───────────────────────── + ───────────────────────── -  ─────────────────────────  for a ≠ -2⎥
⎢f(x) = ⎨                                             , f(x) = ⎨   a⋅log(x)      a⋅log(x)      a⋅log(x)      a⋅log(x)      a⋅log(x)      a⋅log(x)            ⎥
⎢       ⎪               nan                 otherwise          ⎪a⋅ℯ         + 2⋅ℯ           a⋅ℯ         + 2⋅ℯ           a⋅ℯ         + 2⋅ℯ                    ⎥
⎢       ⎩                                                      ⎪                                                                                             ⎥
⎣                                                              ⎩                                       nan                                         otherwise ⎦
```
I'm not sure if this is better than just not solving the equations with more than one symbol with this method until the integral returns the correct result.

Reply:
> I'm not sure if this is better than just not solving the equations with more than one symbol with this method until the integral returns the correct result.

I prefer to have each bit of the codebase that is worked on made correct even if that exposes bugs elsewhere!

Are there not are other cases where this now gives correct/better results?

Reply:
This makes sense.
Yes, this actually helps in some cases. For instance:
```
In [1]: a = Symbol('a', real = True)

In [2]: eq = a*f(x).diff(x) + f(x) + x**2

In [3]: dsolve(eq, hint='1st_exact')
Out[3]:
           -x
           ───
            a
       C₁⋅ℯ         2            2
f(x) = ─────── - 2⋅a  + 2⋅a⋅x - x
          a
```
This one wasn't identified as exact before and now gives a readable solution.
I'll keep looking for the issue with those other examples.

Reply:
> I'll keep looking for the issue with those other examples.

If you push the changes that you currently have up to this PR then I could also take a look.

Reply:
It goes wrong here I think:
https://github.com/sympy/sympy/blob/b76553a1786fcb26eaa0289fa79db4e8d3ce5264/sympy/solvers/ode/ode.py#L2334
The input is constructed as:
```python
from sympy import *

x, a, C1 = symbols('x, a, C1')
f = Function('f')

eq = Eq(Piecewise(((x*exp((a - 1)*log(x)) - exp(a*log(x)))*f(x), (a > -oo) &
    (a < oo) & Ne(a, 0)), ((-a*log(x) + x*exp((a - 1)*log(x)))*f(x), True)) +
    Piecewise((a*f(x)*exp(a*log(x))/(a + 2) + x**2*exp(a*log(x))/(a + 2) +
    2*f(x)*exp(a*log(x))/( a + 2), Ne(a, -2)), (log(x) + f(x)/x**2, True)), C1)

pprint(solve(eq, f(x)))
```
We get:
```python
In [2]: eq
Out[2]: 
                                                                      ⎛⎧        a⋅log(x)    2  a⋅log(x)           a⋅log(x)            ⎞     
                                                                      ⎜⎪a⋅f(x)⋅ℯ           x ⋅ℯ           2⋅f(x)⋅ℯ                    ⎟     
⎛⎧⎛   (a - 1)⋅log(x)    a⋅log(x)⎞                                 ⎞   ⎜⎪──────────────── + ──────────── + ────────────────  for a ≠ -2⎟     
⎜⎪⎝x⋅ℯ               - ℯ        ⎠⋅f(x)  for a > -∞ ∧ a < ∞ ∧ a ≠ 0⎟   ⎜⎪     a + 2            a + 2            a + 2                  ⎟     
⎜⎨                                                                ⎟ + ⎜⎨                                                              ⎟ = C₁
⎜⎪⎛               (a - 1)⋅log(x)⎞                                 ⎟   ⎜⎪                           f(x)                               ⎟     
⎝⎩⎝-a⋅log(x) + x⋅ℯ              ⎠⋅f(x)          otherwise         ⎠   ⎜⎪                  log(x) + ────                     otherwise ⎟     
                                                                      ⎜⎪                             2                                ⎟     
                                                                      ⎝⎩                            x                                 ⎠     

In [3]: solve(eq, f(x))
Out[3]: 
⎡                                    ⎧                            2  a⋅log(x)                          ⎧⎛               2  a⋅log(x)⎞  -a⋅log(x)                                                            ⎤
⎢     ⎧ 2                            ⎪            -C₁⋅a - 2⋅C₁ + x ⋅ℯ                                  ⎪⎝C₁⋅a + 2⋅C₁ - x ⋅ℯ        ⎠⋅ℯ                                                                     ⎥
⎢     ⎪x ⋅(C₁ - log(x))  for a = -2  ⎪────────────────────────────────────────────────────  for a = 0  ⎪───────────────────────────────────────  for (a ≥ -∞ ∧ a < -2) ∨ (a > -2 ∧ a < 0) ∨ (a > 0 ∧ a < ∞)⎥
⎢nan, ⎨                            , ⎨ 2               a⋅log(x)                   a⋅log(x)           , ⎨                 a + 2                                                                             ⎥
⎢     ⎪      nan         otherwise   ⎪a ⋅log(x) - 2⋅a⋅ℯ         + 2⋅a⋅log(x) - 4⋅ℯ                     ⎪                                                                                                   ⎥
⎢     ⎩                              ⎪                                                                 ⎪                  nan                                            otherwise                         ⎥
⎣                                    ⎩                        nan                           otherwise  ⎩                                                                                                   ⎦
```

Reply:
So it's a bug in `solve`.

Reply:
I think that the `solve` bug can be a separate issue. Can you open an issue for that? Then we can get this PR finished and merged.